// @mongodb-js/shell-bson-parser/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParseMode = void 0;
exports.parse = parse;
const acorn_1 = require("acorn");
const check_1 = require("./check");
const eval_1 = require("./eval");
const options_1 = require("./options");
Object.defineProperty(exports, "ParseMode", { enumerable: true, get: function () { return options_1.ParseMode; } });
function buildAST(input) {
    let hasComments = false;
    const ast = (0, acorn_1.parse)(input, {
        ecmaVersion: 6,
        onComment: () => (hasComments = true),
        locations: true,
        ranges: true,
        sourceFile: input,
    });
    return {
        ast,
        hasComments,
    };
}
function parse(input, options) {
    const parsedOptions = (0, options_1.buildOptions)(options);
    const { hasComments, ast } = buildAST(`(\n${input}\n)`);
    const passedCommentsCheck = !hasComments || parsedOptions.allowComments;
    if (passedCommentsCheck && (0, check_1.checkTree)(ast, parsedOptions)) {
        return (0, eval_1.executeAST)(ast);
    }
    return '';
}
exports.default = parse;
