// @mongodb-js/shell-bson-parser/dist/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParseMode = void 0;
exports.buildOptions = buildOptions;
var ParseMode;
(function (ParseMode) {
    ParseMode["Strict"] = "strict";
    ParseMode["Extended"] = "extended";
    ParseMode["Loose"] = "loose";
})(ParseMode || (exports.ParseMode = ParseMode = {}));
const StrictOptions = {
    allowMethods: false,
    allowComments: false,
};
const ExtendedOptions = {
    allowMethods: true,
};
const LooseOptions = {
    allowMethods: true,
    allowComments: true,
};
function getModeOptions(mode) {
    switch (mode) {
        case ParseMode.Strict:
            return StrictOptions;
        case ParseMode.Extended:
            return ExtendedOptions;
        case ParseMode.Loose:
            return LooseOptions;
    }
}
const DefaultOptions = {
    mode: ParseMode.Strict,
    ...StrictOptions,
};
function buildOptions(options) {
    return {
        ...DefaultOptions,
        ...getModeOptions((options && options.mode) || ParseMode.Strict),
        ...options,
    };
}
