// mongodb-query-parser/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_MAX_TIME_MS = exports.DEFAULT_COLLATION = exports.DEFAULT_PROJECT = exports.DEFAULT_SKIP = exports.DEFAULT_LIMIT = exports.DEFAULT_SORT = exports.DEFAULT_FILTER = exports.QUERY_PROPERTIES = exports.toJSString = exports.stringify = void 0;
exports.parseSort = parseSort;
exports.parseFilter = parseFilter;
exports.parseCollation = parseCollation;
exports.isFilterValid = isFilterValid;
exports.isCollationValid = isCollationValid;
exports.parseProject = parseProject;
exports.isProjectValid = isProjectValid;
exports.isSortValid = isSortValid;
exports.isMaxTimeMSValid = isMaxTimeMSValid;
exports.isSkipValid = isSkipValid;
exports.isLimitValid = isLimitValid;
exports.validate = validate;
exports.default = queryParser;
const shell_bson_parser_1 = __importStar(require("@mongodb-js/shell-bson-parser"));
const lodash_1 = __importDefault(require("lodash"));
const debug_1 = __importDefault(require("debug"));
const constants_1 = require("./constants");
const stringify_1 = require("./stringify");
Object.defineProperty(exports, "stringify", { enumerable: true, get: function () { return stringify_1.stringify; } });
Object.defineProperty(exports, "toJSString", { enumerable: true, get: function () { return stringify_1.toJSString; } });
const debug = (0, debug_1.default)('mongodb-query-parser');
const DEFAULT_FILTER = {};
exports.DEFAULT_FILTER = DEFAULT_FILTER;
const DEFAULT_SORT = null;
exports.DEFAULT_SORT = DEFAULT_SORT;
const DEFAULT_LIMIT = 0;
exports.DEFAULT_LIMIT = DEFAULT_LIMIT;
const DEFAULT_SKIP = 0;
exports.DEFAULT_SKIP = DEFAULT_SKIP;
const DEFAULT_PROJECT = null;
exports.DEFAULT_PROJECT = DEFAULT_PROJECT;
const DEFAULT_COLLATION = null;
exports.DEFAULT_COLLATION = DEFAULT_COLLATION;
const DEFAULT_MAX_TIME_MS = 60000;
exports.DEFAULT_MAX_TIME_MS = DEFAULT_MAX_TIME_MS;
const QUERY_PROPERTIES = ['filter', 'project', 'sort', 'skip', 'limit'];
exports.QUERY_PROPERTIES = QUERY_PROPERTIES;
function isEmpty(input) {
    if (input === null || input === undefined) {
        return true;
    }
    const s = lodash_1.default.trim(typeof input === 'number' ? `${input}` : input);
    if (s === '{}') {
        return true;
    }
    return lodash_1.default.isEmpty(s);
}
function isNumberValid(input) {
    if (isEmpty(input)) {
        return 0;
    }
    return /^\d+$/.test(`${input}`) ? parseInt(`${input}`, 10) : false;
}
function _parseProject(input) {
    return (0, shell_bson_parser_1.default)(input, { mode: shell_bson_parser_1.ParseMode.Loose });
}
function _parseCollation(input) {
    return (0, shell_bson_parser_1.default)(input, { mode: shell_bson_parser_1.ParseMode.Loose });
}
function parseSort(input) {
    if (isEmpty(input)) {
        return DEFAULT_SORT;
    }
    return (0, shell_bson_parser_1.default)(input, { mode: shell_bson_parser_1.ParseMode.Loose });
}
function _parseFilter(input) {
    return (0, shell_bson_parser_1.default)(input, {
        mode: shell_bson_parser_1.ParseMode.Loose,
        allowMethods: true,
    });
}
function parseFilter(input) {
    if (isEmpty(input)) {
        return DEFAULT_FILTER;
    }
    return _parseFilter(input);
}
function parseCollation(input) {
    if (isEmpty(input)) {
        return DEFAULT_COLLATION;
    }
    return _parseCollation(input);
}
function isFilterValid(input) {
    if (isEmpty(input)) {
        return DEFAULT_FILTER;
    }
    try {
        return _parseFilter(input);
    }
    catch (e) {
        debug('Filter "%s" is invalid', input, e);
        return false;
    }
}
function _isCollationValid(collation) {
    for (const [key, value] of Object.entries(collation)) {
        if (!constants_1.COLLATION_OPTIONS[key]) {
            debug('Collation "%s" is invalid bc of its keys', collation);
            return false;
        }
        if (constants_1.COLLATION_OPTIONS[key].includes(value) === false) {
            debug('Collation "%s" is invalid bc of its values', collation);
            return false;
        }
    }
    return collation;
}
function isCollationValid(input) {
    if (isEmpty(input)) {
        return DEFAULT_COLLATION;
    }
    try {
        const parsed = _parseCollation(input);
        return _isCollationValid(parsed);
    }
    catch (e) {
        debug('Collation "%s" is invalid', input, e);
        return false;
    }
}
function isValueOkForProject() {
    return true;
}
function parseProject(input) {
    if (isEmpty(input)) {
        return DEFAULT_PROJECT;
    }
    return _parseProject(input);
}
function isProjectValid(input) {
    if (isEmpty(input)) {
        return DEFAULT_PROJECT;
    }
    try {
        const parsed = _parseProject(input);
        if (!lodash_1.default.isObject(parsed)) {
            debug('Project "%s" is invalid. Only documents are allowed', input);
            return false;
        }
        if (!lodash_1.default.every(parsed, isValueOkForProject)) {
            debug('Project "%s" is invalid bc of its values', input);
            return false;
        }
        return parsed;
    }
    catch (e) {
        debug('Project "%s" is invalid', input, e);
        return false;
    }
}
const ALLOWED_SORT_VALUES = [1, -1, 'asc', 'desc'];
function isValueOkForSortDocument(val) {
    return (lodash_1.default.includes(ALLOWED_SORT_VALUES, val) ||
        !!(lodash_1.default.isObject(val) && val.$meta));
}
function isValueOkForSortArray(val) {
    return (lodash_1.default.isArray(val) &&
        val.length === 2 &&
        lodash_1.default.isString(val[0]) &&
        isValueOkForSortDocument(val[1]));
}
function isSortValid(input) {
    try {
        const parsed = parseSort(input);
        if (isEmpty(parsed)) {
            return DEFAULT_SORT;
        }
        if (lodash_1.default.isArray(parsed) && lodash_1.default.every(parsed, isValueOkForSortArray)) {
            return parsed;
        }
        if (lodash_1.default.isObject(parsed) &&
            !lodash_1.default.isArray(parsed) &&
            lodash_1.default.every(parsed, isValueOkForSortDocument)) {
            return parsed;
        }
        debug('Sort "%s" is invalid bc of its values', input);
        return false;
    }
    catch (e) {
        debug('Sort "%s" is invalid', input, e);
        return false;
    }
}
function isMaxTimeMSValid(input) {
    if (isEmpty(input)) {
        return DEFAULT_MAX_TIME_MS;
    }
    return isNumberValid(input);
}
function isSkipValid(input) {
    if (isEmpty(input)) {
        return DEFAULT_SKIP;
    }
    return isNumberValid(input);
}
function isLimitValid(input) {
    if (isEmpty(input)) {
        return DEFAULT_LIMIT;
    }
    return isNumberValid(input);
}
const validatorFunctions = {
    isMaxTimeMSValid,
    isFilterValid,
    isProjectValid,
    isSortValid,
    isLimitValid,
    isSkipValid,
    isCollationValid,
    isNumberValid,
};
function validate(what, input) {
    const validator = validatorFunctions[`is${lodash_1.default.upperFirst(what)}Valid`];
    if (!validator) {
        debug('Do not know how to validate `%s`. Returning false.', what);
        return false;
    }
    return validator(input);
}
function queryParser(filter, project = DEFAULT_PROJECT) {
    if (arguments.length === 1) {
        if (lodash_1.default.isString(filter)) {
            return _parseFilter(filter);
        }
    }
    return {
        filter: _parseFilter(filter),
        project: project !== DEFAULT_PROJECT ? _parseProject(project) : project,
    };
}
